src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/CoverageModelEMWorkspace.java

```diff
@@ -0,0 +1,2259 @@
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.api.java.function.Function2;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.storage.StorageLevel;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.tools.coveragemodel.annotations.CachesRDD;
+import org.broadinstitute.hellbender.tools.coveragemodel.annotations.EvaluatesRDD;
+import org.broadinstitute.hellbender.tools.coveragemodel.annotations.UpdatesRDD;
+import org.broadinstitute.hellbender.tools.coveragemodel.interfaces.CopyRatioExpectationsCalculator;
+import org.broadinstitute.hellbender.tools.coveragemodel.linalg.FourierLinearOperator;
+import org.broadinstitute.hellbender.tools.coveragemodel.linalg.FourierLinearOperatorNDArray;
+import org.broadinstitute.hellbender.tools.coveragemodel.linalg.GeneralLinearOperator;
+import org.broadinstitute.hellbender.tools.coveragemodel.linalg.IterativeLinearSolverNDArray;
+import org.broadinstitute.hellbender.tools.coveragemodel.linalg.IterativeLinearSolverNDArray.ExitStatus;
+import org.broadinstitute.hellbender.tools.coveragemodel.math.RobustBrentSolver;
+import org.broadinstitute.hellbender.tools.coveragemodel.math.SynchronizedUnivariateSolver;
+import org.broadinstitute.hellbender.tools.coveragemodel.nd4jutils.Nd4jIOUtils;
+import org.broadinstitute.hellbender.tools.exome.*;
+import org.broadinstitute.hellbender.tools.exome.sexgenotyper.GermlinePloidyAnnotatedTargetCollection;
+import org.broadinstitute.hellbender.tools.exome.sexgenotyper.SexGenotypeData;
+import org.broadinstitute.hellbender.tools.exome.sexgenotyper.SexGenotypeDataCollection;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.hmm.interfaces.AlleleMetadataProducer;
+import org.broadinstitute.hellbender.utils.hmm.interfaces.CallStringProducer;
+import org.broadinstitute.hellbender.utils.hmm.interfaces.ScalarProducer;
+import org.broadinstitute.hellbender.utils.param.ParamUtils;
+import org.nd4j.linalg.api.buffer.DataBuffer;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.factory.Nd4j;
+import org.nd4j.linalg.indexing.INDArrayIndex;
+import org.nd4j.linalg.indexing.NDArrayIndex;
+import org.nd4j.linalg.ops.transforms.Transforms;
+import scala.Tuple2;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static org.broadinstitute.hellbender.tools.coveragemodel.CoverageModelEMParams.CopyRatioHMMType.COPY_RATIO_HMM_LOCAL;
+
+/**
+ * This class represents the driver-node workspace for EM algorithm calculations of the coverage model
```

Comment:
Too much of the code is publicly exposed at the moment. I will make an issue for that.